extensions/redis-cache/src/main/java/io/druid/client/cache/RedisCache.java

```diff
@@ -0,0 +1,288 @@
+  private final AtomicLong hitCount = new AtomicLong(0);
+  private final AtomicLong missCount = new AtomicLong(0);
+  private final AtomicLong timeoutCount = new AtomicLong(0);
+  private final AtomicLong errorCount = new AtomicLong(0);
+
+  public static Cache create(final RedisCacheConfig config)
+  {
+    JedisPoolConfig poolConfig = config.getPoolConfig();
+    JedisPool pool = new JedisPool(poolConfig, config.getHost(), config.getPort(), config.getTimeout());
+    return new RedisCache(pool, config);
+  }
+
+  public RedisCache(JedisPool pool, RedisCacheConfig config)
+  {
+    this.pool = pool;
+    prefix = config.getPrefix();
+    expiration = config.getExpiration();
+    transcoder = new RedisTranscoder();
+  }
+
+  @Override
+  public byte[] get(NamedKey key)
+  {
+    final byte[] hash = computeKeyHash(prefix, key);
+    byte[] result = null;
+
+    try (final Jedis client = pool.getResource()) {
+      final byte[] serialized = client.get(hash);
+
+      if (serialized != null) {
+        result = deserialize(key, serialized);
+        hitCount.incrementAndGet();
+      }
+      else {
+        missCount.incrementAndGet();
+      }
+    }
+    catch (JedisConnectionException ex) {
+      if (isTimeout(ex)) {
+        timeoutCount.incrementAndGet();
+      }
+
+      throw Throwables.propagate(ex);
+    }
+    catch (JedisException | LZ4Exception ex) {
+      errorCount.incrementAndGet();
+      log.warn(ex, "Error getting cache item");
+      return null;
+    }
+
+    return result;
+  }
+
+  @Override
+  public void put(NamedKey key, byte[] value)
+  {
+    final byte[] hash = computeKeyHash(prefix, key);
+    final byte[] serialized;
+
+    try {
+      serialized = serialize(key, value);
+    }
+    catch (NullPointerException ex) {
+      throw Throwables.propagate(ex);
+    }
+
+    try (final Jedis client = pool.getResource()) {
+      client.psetex(hash, expiration, serialized);
+    }
+    catch (JedisConnectionException ex) {
+      Throwables.propagate(ex);
+    }
+    catch (JedisException ex) {
+      errorCount.incrementAndGet();
+      log.warn(ex, "Error putting cache item");
+    }
+  }
+
+  @Override
+  public Map<NamedKey, byte[]> getBulk(Iterable<NamedKey> keys)
+  {
+    final Map<byte[], NamedKey> keyLookup = Maps.uniqueIndex(
+      keys,
+      new Function<NamedKey, byte[]>()
+      {
+        @Override
+        public byte[] apply(
+          @Nullable NamedKey input
+        )
+        {
+          return computeKeyHash(prefix, input);
+        }
+      }
+    );
+    final Map<NamedKey, byte[]> results = Maps.newHashMap();
+
+    // Sometimes broker passes empty keys list to getBulk()
+    if (keyLookup.size() == 0) {
+      return results;
+    }
+
+    final List<byte[]> keysList = new ArrayList<>(keyLookup.keySet());
+    final List<byte[]> valuesList;
+
+    try (final Jedis client = pool.getResource()) {
+      valuesList = client.mget(keysList.toArray(new byte[keysList.size()][]));
```

Comment:
how about instead of creating a separate array of keys, you created keyLookup as a LinkedHashMap, this way you can simply call `Iterables.toArray(keyLookup.keySet())`